crypto/digestible/derive/test/tests/behavior.rs

```diff
@@ -67,38 +42,81 @@ struct GenericFoo<X: Digestible> {
 #[test]
 fn foo1() {
     let arg = Foo { a: 0, b: 1, c: 2 };
-    let mut hasher = Tester::new();
-    arg.digest(&mut hasher);
-
-    let expected: Vec<Vec<u8>> = vec![
-        b"Foo".to_vec(),
-        b"".to_vec(),
-        b"a".to_vec(),
-        vec![0u8, 0u8],
-        b"b".to_vec(),
-        vec![1u8, 0u8],
-        b"c".to_vec(),
-        vec![2u8, 0u8],
-    ];
-
-    assert_eq!(hasher.args, expected);
+    let expected_ast = ASTNode::from(ASTAggregate {
+        context: b"foo1",
+        name: b"Foo".to_vec(),
+        elems: vec![
+            ASTNode::from(ASTPrimitive {
+                context: b"a",
+                type_name: b"uint",
+                data: vec![0u8, 0u8],
+            }),
+            ASTNode::from(ASTPrimitive {
+                context: b"b",
+                type_name: b"uint",
+                data: vec![1u8, 0u8],
+            }),
+            ASTNode::from(ASTPrimitive {
+                context: b"c",
+                type_name: b"uint",
+                data: vec![2u8, 0u8],
+            }),
+        ],
+        is_completed: true,
+    });
+    digestible_test_case_ast("foo1", &arg, expected_ast);
+    assert_eq!(
+        arg.digest32::<MerlinTranscript>(b"foo1"),
+        [
+            44, 174, 200, 201, 207, 72, 151, 193, 63, 97, 49, 112, 135, 50, 252, 61, 211, 208, 41,
+            4, 241, 249, 90, 114, 181, 197, 177, 151, 39, 165, 76, 197
+        ]
+    );
 }
 
 #[test]
 fn blob1() {
     let arg = Blob(vec![1, 2, 3, 4]);
-    let mut hasher = Tester::new();
-    arg.digest(&mut hasher);
-
-    let expected: Vec<Vec<u8>> = vec![
-        b"Blob".to_vec(),
-        b"".to_vec(),
-        b"0".to_vec(),
-        4usize.to_le_bytes().to_vec(),
-        vec![1u8, 2u8, 3u8, 4u8],
-    ];
-
-    assert_eq!(hasher.args, expected);
+    let expected_ast = ASTNode::Aggregate(ASTAggregate {
+        context: b"blob1",
+        name: b"Blob".to_vec(),
+        elems: vec![ASTNode::Primitive(ASTPrimitive {
+            context: b"0",
+            type_name: b"bytes",
+            data: vec![1u8, 2u8, 3u8, 4u8],
+        })],
+        is_completed: true,
+    });
+    digestible_test_case_ast("blob1", &arg, expected_ast);
+    assert_eq!(
+        arg.digest32::<MerlinTranscript>(b"blob1"),
+        [
+            59, 63, 205, 99, 4, 221, 221, 230, 90, 1, 135, 226, 106, 52, 210, 105, 88, 37, 182, 26,
+            208, 240, 152, 4, 226, 0, 204, 11, 10, 187, 14, 48
+        ]
+    );
+}
+
+#[test]
+fn blob2() {
```

Comment:
Again, it would help if the function name and/or comment describe what's happening here, e.g. "A tuple struct with the `transparent` attribute is not treated as an aggregate. This is intended to work with the "new type" pattern of wrapping an existing data type."

Also,  does #[digestible(transparent)] work for tuple structs with more than one field? Or structs with named fields? Or unit (fieldless) structs?

Reply:
the transparent thing only works when it has exactly one field, it's meant for like:

you have a type that's in the ledger somehow, and you need to create a newtype wrapper around it, but you don't want that wrapper to change the hash

i added `transparent` to a bunch of wrappers in the transaction crate in this commit. i'm not sure if that was a great idea -- it will mean that we can't add fields to those types. but mainly they are like "TxOutMembershipHash" and its hard to imagine that we would add fields to them. the main benefit of marking them transparent was that it made the AST things more readable for important structures like TxOut. but possibly we shouldn't actually make those transparent.

I also made `RistrettoPublic` transparent, because I don't think we want to add fields to it, and that made the AST more readable. but at the end of the day, it doesn't matter that much if the AST is readable, and its concievable, though unlikely that we might want to add fields to that thing. so that's kind of a difficult judgment call

Reply:
That makes sense, and that is a tricky judgement call.

I like that `transparent` lets us refactor a field to be a newtype (or a different newtype with the same inner data) without changing the digest. That might be handy if we decided that, say, RistrettoPublic wasn't specific enough, and we wanted different newtypes to denote different categories of Ristretto public keys.